Objects/setobject.c

```diff
@@ -1063,36 +1063,73 @@ make_new_set_basetype(PyTypeObject *type, PyObject *iterable)
 /* The empty frozenset is a singleton */
 static PyObject *emptyfrozenset = NULL;
 
+static PyObject *
+make_new_frozenset(PyTypeObject *type, PyObject *iterable)
+{
+    if (iterable != NULL) {
+        /* frozenset(f) is idempotent */
+        if (PyFrozenSet_CheckExact(iterable)) {
+            Py_INCREF(iterable);
+            return iterable;
+        }
+        PyObject *res = make_new_set((PyTypeObject *)type, iterable);
+        if (res == NULL || PySet_GET_SIZE(res)) {
+            return res;
+        }
+        Py_DECREF(res);
+    }
+
+    // The empty frozenset is a singleton
+    if (emptyfrozenset == NULL) {
+        emptyfrozenset = make_new_set((PyTypeObject *)type, NULL);
+    }
+    Py_XINCREF(emptyfrozenset);
+    return emptyfrozenset;
+}
+
 static PyObject *
 frozenset_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
 {
-    PyObject *iterable = NULL, *result;
+    PyObject *iterable = NULL;
 
-    if (type == &PyFrozenSet_Type && !_PyArg_NoKeywords("frozenset", kwds))
+    if (type == &PyFrozenSet_Type && !_PyArg_NoKeywords("frozenset", kwds)) {
         return NULL;
+    }
 
-    if (!PyArg_UnpackTuple(args, type->tp_name, 0, 1, &iterable))
+    if (!PyArg_UnpackTuple(args, type->tp_name, 0, 1, &iterable)) {
         return NULL;
+    }
 
-    if (type != &PyFrozenSet_Type)
+    if (type != &PyFrozenSet_Type) {
         return make_new_set(type, iterable);
+    }
 
     if (iterable != NULL) {
-        /* frozenset(f) is idempotent */
-        if (PyFrozenSet_CheckExact(iterable)) {
-            Py_INCREF(iterable);
-            return iterable;
-        }
-        result = make_new_set(type, iterable);
-        if (result == NULL || PySet_GET_SIZE(result))
-            return result;
-        Py_DECREF(result);
+        return make_new_frozenset(type, iterable);
     }
-    /* The empty frozenset is a singleton */
-    if (emptyfrozenset == NULL)
-        emptyfrozenset = make_new_set(type, NULL);
-    Py_XINCREF(emptyfrozenset);
-    return emptyfrozenset;
+    return make_new_frozenset(type, NULL);
+}
+
+static PyObject *
+frozenset_vectorcall(PyObject *type, PyObject * const*args,
+                     size_t nargsf, PyObject *kwnames)
+{
+    assert(PyType_Check(type));
+
+    if (!_PyArg_NoKwnames("frozenset", kwnames)) {
+        return NULL;
+    }
+
+    Py_ssize_t nargs = PyVectorcall_NARGS(nargsf);
+    if (!_PyArg_CheckPositional("frozenset", nargs, 0, 1)) {
+        return NULL;
+    }
+
+    if (nargs) {
+        return make_new_frozenset((PyTypeObject *)type, args[0]);
+    }
+
+    return make_new_frozenset((PyTypeObject *)type, NULL);
```

Comment:
You may replace these 3 lines with:

```suggestion
    PyObject *iterable = (nargs ? args[0] : NULL);
    return make_new_frozenset((PyTypeObject *)type, iterable);
```